### PR TITLE
Generate thumbnail images for video uploads

### PR DESCRIPTION
**Pull request by oblakeerickson**
FEATURE: Generate thumbnail images for uploaded videos

Topics in Discourse have a topic thumbnail feature which allows themes
to show a preview image before viewing the actual Topic.

This PR allows for the ability to generate a thumbnail image from an
uploaded video that can be use for the topic preview.

### CODE REVIEW

File: spec/lib/cooked_post_processor_spec.rb [hunk 1]
```diff
@@ -1456,6 +1456,28 @@
           HTML
         end
 
+        context "with video upload" do
+          SiteSetting.authorized_extensions = "png|jpg|gif|mp4"
```
Comment (jjaffeux) [inline spec/lib/cooked_post_processor_spec.rb:1460]: Should this be in a before block? Also you probably just need: `SiteSetting.authorized_extensions = "mp4"`

File: app/models/post.rb [hunk 2]
```diff
@@ -1007,6 +1007,25 @@ def link_post_uploads(fragments: nil)
       upload = nil
       upload = Upload.find_by(sha1: sha1) if sha1.present?
       upload ||= Upload.get_from_url(src)
+
+      # Link any video thumbnails
+      if upload.present?
+        # Video thumbnails have the filename of the video file sha1 with a .png or .jpg extention.
```
Comment (pmusaraj) [inline app/models/post.rb:1013]:
```suggestion
        # Video thumbnails have the filename of the video file sha1 with a .png or .jpg extension.
```

This is interesting. Maybe we can restrict this to video-type uploads only?

File: spec/lib/cooked_post_processor_spec.rb [hunk 3]
```diff
@@ -1568,6 +1568,26 @@
...
+
+          it "works" do
+            post = Fabricate(:post, raw: "test video upload:\n\n![test|video](#{video_upload.url})")
+            cpp = CookedPostProcessor.new(post.reload)
+            expect(cpp.html).to match_html <<~HTML
+              <p>test video upload:</p>
+              <p></p><div class="video-container">
+                <video width="100%" height="100%" preload="metadata" controls="">
+                  <source src="/uploads/default/test_0/#{Discourse.store.get_path_for_upload(video_upload)}">
+                  <a href="/uploads/default/test_0/#{Discourse.store.get_path_for_upload(video_upload)}">/uploads/default/test_0/#{Discourse.store.get_path_for_upload(video_upload)}</a>
```
Comment (pmusaraj) [inline spec/lib/cooked_post_processor_spec.rb:1584]: Is this testing anything related to video thumbnails?

File: app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    let videoHeight, videoWidth;
+    videoHeight = video.videoHeight;
+    videoWidth = video.videoWidth;
+    canvas.width = videoWidth;
+    canvas.height = videoHeight;
+
+    ctx.drawImage(video, 0, 0, videoWidth, videoHeight);
+
+    // upload video thumbnail
+    canvas.toBlob((blob) => {
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js:45]: I pulled this branch locally, and if the image URL and site domain don't match, I get this error: 

```
Failed to execute 'toBlob' on 'HTMLCanvasElement': Tainted canvases may not be exported.
```

Locally I get this for `localhost` vs `127.0.0.1` but in prod this can happen, too, maybe?

Comment (oblakeerickson) [reply app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js:45]: Ya, locally it has to be 127.0.0.1. I don't think this would happen in prod. I think this is just related to our ember_cli proxy. We do use a some other canvas exporting related things in core already that work in prod:

```
app/assets/javascripts/discourse/app/lib/update-tab-count.js
49:  newFavicon.href = canvas.toDataURL("image/png");
```

File: app/assets/javascripts/discourse/app/components/composer-editor.js [hunk 5]
```diff
@@ -98,453 +99,447 @@ export function addComposerUploadMarkdownResolver(resolver) {
 export function cleanUpComposerUploadMarkdownResolver() {
   uploadMarkdownResolvers = [];
 }
+export default Component.extend(
+  ComposerUploadUppy,
+  ComposerVideoThumbnailUppy,
```
Comment (oblakeerickson) [inline app/assets/javascripts/discourse/app/components/composer-editor.js:104]: Most changes are prettier indentation changes caused by adding this line. Try viewing without whitespace changes to see a better diff https://github.com/discourse/discourse/pull/19801/files?w=1

File: app/assets/javascripts/discourse/app/components/composer-editor.js [hunk 6]
```diff
@@ -554,374 +549,402 @@ export default Component.extend(ComposerUploadUppy, {
...
+            post.set("refreshedPost", true);
+          }
+        };
+
+        discourseDebounce(this, paintFunc, 450);
+
+        // Short upload urls need resolution
+        resolveAllShortUrls(ajax, this.siteSettings, preview);
+
+        this._generateVideoThumbnail();
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/composer-editor.js:940]: Since this only works when diff preview is enabled, should we explicitly only run thumbnail generation for videos only if that setting is checked?

File: app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+          this.set("uploading", false);
+        });
+
+        this._uppyInstance.on("upload-error", (file, error, response) => {
+          let message = I18n.t("wizard.upload_error");
+          if (response.body.errors) {
+            message = response.body.errors.join("\n");
+          }
+
+          this.dialog.alert(message);
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js:89]: I'm not sure we should raise an alert here for users. A console error might be good enough (and less intrusive).

File: app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+  _useXHRUploads() {
+    this._uppyInstance.use(XHRUpload, {
+      endpoint: getURL(`/uploads.json?client_id=${this.messageBus.clientId}`),
+      headers: {
+        "X-CSRF-Token": this.session.csrfToken,
+      },
+    });
+  },
+});
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js:122]: Is it possible to test this somewhat? There is a bit of frontend/backend interaction here which makes it tricky, but a useful test would be to have a frontend test of a video file in the composer that triggers an uppy upload. 

Or, possibly a system test, the tricky thing there will be to find a way to display the thumbnail somewhere, I'm not sure we can easily do that in core (without theme components or plugins). 

(That said, given this feature is experimental, I'm OK with handling the test in a followup PR.)

**Review by pmusaraj — APPROVED** [review]
Looks generally good to me, left a few minor comments, but approving, thanks!

File: app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js (hunk 8 above)
Comment (oblakeerickson) [reply app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js:122]: > (That said, given this feature is experimental, I'm OK with handling the test in a followup PR.)

:+1: I'm going to hold off for now then. I definitely want an end-to-end test, just not sure it's worth the effort at this stage.
